### PR TITLE
Symfony3 support (merged forks)

### PR DESCRIPTION
**Pull request by ownede**
Merged my, @ubermuda, @bassrock and @DenisMedved code together. 

As mentioned in #275 basic features seems to be working, no more deprecated code.
> Creating a new thread, replying, views and forms (outbox/inbox/deleted/new message/reply) seems to work great. Need some more time for testing.

Currently testing this on my project. 
Thanks for anyone who contributed.

### CODE REVIEW

**Comment by tgalopin** [comment]
That's an awesome work, I'll check this tonight! Thanks!

**Comment by ubermuda** [comment]
Amazing work! Thanks for your efforts.

**Comment by Filoz** [comment]
Thanks for your work!

Will this be merge?

File: DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -51,7 +51,7 @@ public function getConfigTreeBuilder()
                     ->addDefaultsIfNotSet()
                     ->children()
                         ->scalarNode('factory')->defaultValue('fos_message.new_thread_form.factory.default')->cannotBeEmpty()->end()
-                        ->scalarNode('type')->defaultValue('fos_message.new_thread_form.type.default')->cannotBeEmpty()->end()
+                        ->scalarNode('class')->defaultValue('FOS\MessageBundle\FormType\NewThreadMessageFormType')->cannotBeEmpty()->end()
```
Comment (Ener-Getick) [inline DependencyInjection/Configuration.php]: this is clearly a bc break, we need a bc layer to keep supporting the old option

File: FormFactory/AbstractMessageFormFactory.php [hunk 2]
```diff
@@ -21,11 +20,11 @@
...
 
     /**
-     * The message form type
+     * The message form class
      *
-     * @var AbstractType
+     * @var string
      */
-    protected $formType;
+    protected $formClass;
```
Comment (Ener-Getick) [inline FormFactory/AbstractMessageFormFactory.php]: that's a bc break as well

File: FormHandler/AbstractMessageFormHandler.php [hunk 3]
```diff
@@ -22,9 +23,9 @@
     protected $sender;
     protected $participantProvider;
 
-    public function __construct(Request $request, ComposerInterface $composer, SenderInterface $sender, ParticipantProviderInterface $participantProvider)
+    public function __construct(RequestStack $requestStack, ComposerInterface $composer, SenderInterface $sender, ParticipantProviderInterface $participantProvider)
```
Comment (Ener-Getick) [inline FormHandler/AbstractMessageFormHandler.php]: bc break

File: FormHandler/AbstractMessageFormHandler.php [hunk 4]
```diff
@@ -22,9 +23,9 @@
...
     {
-        $this->request = $request;
+        $this->request = $requestStack->getCurrentRequest();
```
Comment (Ener-Getick) [inline FormHandler/AbstractMessageFormHandler.php]: The request should not be retrieved in the constructor.
You can do something like:
```php
/**
 * @param Request|RequestStack $request
 */
public function __construct($request, ...) {
    if ($request instanceof Request) {
         @trigger_error(sprintf('Using an instance of "%s" as first parameter of "%s" is deprecated since version 1.3 and won\'t be supported in 2.0. Use an instance of "Symfony\Component\HttpFoundation\RequestStack" instead.', get_class($request), __METHOD), E_USER_DEPRECATED);
    }
    $this->request = $request;
    // ...
}

private function getCurrentRequest() {
   if ($this->request instanceof Request) {
      return $this->request;
   }
   
   return $this->request->getCurrentRequest();
}

public function process() {
    // ...
    $request = $this->getCurrentRequest();
}

File: FormType/NewThreadMessageFormType.php [hunk 5]
```diff
@@ -16,19 +19,19 @@ class NewThreadMessageFormType extends AbstractType
...
-            ->add('recipient', 'fos_user_username', array('label' => 'recipient', 'translation_domain' => 'FOSMessageBundle'))
-            ->add('subject', 'text', array('label' => 'subject', 'translation_domain' => 'FOSMessageBundle'))
-            ->add('body', 'textarea', array('label' => 'body', 'translation_domain' => 'FOSMessageBundle'));
+            ->add('recipient', UsernameFormType::class, array('label' => 'recipient', 'translation_domain' => 'FOSMessageBundle'))
+            ->add('subject', TextType::class, array('label' => 'subject', 'translation_domain' => 'FOSMessageBundle'))
+            ->add('body', TextareaType::class, array('label' => 'body', 'translation_domain' => 'FOSMessageBundle'));
     }
 
-    public function setDefaultOptions(OptionsResolverInterface $resolver)
+    public function configureOptions(OptionsResolver $resolver)
```
Comment (Ener-Getick) [inline FormType/NewThreadMessageFormType.php]: thats a bc break.

please replace that by:
```php
public function configureOptions(OptionsResolver $resolver)
{
    $resolver->setDefaults(array(
        'intention'  => 'message',
     ));
}

public function setDefaultOptions(OptionsResolverInterface $resolver)
{
    $resolver->setDefaults(array(
        'intention'  => 'message',
     ));
}

File: FormType/NewThreadMessageFormType.php [hunk 6]
```diff
@@ -16,19 +19,19 @@ class NewThreadMessageFormType extends AbstractType
...
     {
         $resolver->setDefaults(array(
             'intention'  => 'message',
         ));
     }
 
-    public function getName()
+    public function getBlockPrefix()
```
Comment (Ener-Getick) [inline FormType/NewThreadMessageFormType.php]: same here, this function should be an alias of ``getName`` for bc

File: FormType/NewThreadMultipleMessageFormType.php [hunk 7]
```diff
@@ -14,12 +16,12 @@ class NewThreadMultipleMessageFormType extends AbstractType
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
         $builder
-            ->add('recipients', 'recipients_selector', array('label' => 'recipients', 'translation_domain' => 'FOSMessageBundle'))
-            ->add('subject', 'text', array('label' => 'subject', 'translation_domain' => 'FOSMessageBundle'))
-            ->add('body', 'textarea', array('label' => 'body', 'translation_domain' => 'FOSMessageBundle'));
+            ->add('recipients', RecipientsType::class, array('label' => 'recipients', 'translation_domain' => 'FOSMessageBundle'))
```
Comment (Ener-Getick) [inline FormType/NewThreadMultipleMessageFormType.php]: Should be replaced by something like https://github.com/FriendsOfSymfony/FOSUserBundle/blob/master/Util/LegacyFormHelper.php

File: FormType/NewThreadMultipleMessageFormType.php [hunk 8]
```diff
@@ -14,12 +16,12 @@ class NewThreadMultipleMessageFormType extends AbstractType
...
+            ->add('subject', TextType::class, array('label' => 'subject', 'translation_domain' => 'FOSMessageBundle'))
+            ->add('body', TextareaType::class, array('label' => 'body', 'translation_domain' => 'FOSMessageBundle'));
     }
 
-    public function getName()
+    public function getBlockPrefix()
```
Comment (Ener-Getick) [inline FormType/NewThreadMultipleMessageFormType.php]: should alias ``getName``

File: FormType/RecipientsType.php [hunk 9]
```diff
@@ -50,13 +51,13 @@ public function setDefaultOptions(OptionsResolverInterface $resolver)
      */
     public function getParent()
     {
-        return 'text';
+        return TextType::class;
```
Comment (Ener-Getick) [inline FormType/RecipientsType.php]: ``LegacyFormHelper::isLegacy() ? 'text' : 'Symfony\Component\Form\Extension\Core\Type\TextType';`` we can't use short notation yet.

File: Resources/config/config.xml [hunk 10]
```diff
@@ -58,8 +58,8 @@
             <tag name="twig.extension" alias="fos_message" />
         </service>
 
-        <service id="fos_message.search_query_factory.default" class="FOS\MessageBundle\Search\QueryFactory" scope="request" public="true">
-            <argument type="service" id="request" />
+        <service id="fos_message.search_query_factory.default" class="FOS\MessageBundle\Search\QueryFactory" public="true">
+            <argument type="service" id="request_stack" />
```
Comment (Ener-Getick) [inline Resources/config/config.xml]: this should be managed in a compiler pass or we must remove support of ``symfony/framework-bundle < 2.4``.

File: Security/ParticipantProvider.php [hunk 11]
```diff
@@ -14,15 +14,15 @@
...
+     * The token storage
      *
-     * @var SecurityContextInterface
+     * @var TokenStorageInterface
      */
-    protected $securityContext;
+    protected $tokenStorage;
 
-    public function __construct(SecurityContextInterface $securityContext)
+    public function __construct(TokenStorageInterface $tokenStorage)
```
Comment (Ener-Getick) [inline Security/ParticipantProvider.php]: That's a bc break, the class should support both implementations.

File: Search/QueryFactory.php [hunk 12]
```diff
@@ -24,12 +25,12 @@ class QueryFactory implements QueryFactoryInterface
...
      *
-     * @param Request $request
+     * @param RequestStack $requestStack
      * @param string $queryParameter
      */
-    public function __construct(Request $request, $queryParameter)
+    public function __construct(RequestStack $requestStack, $queryParameter)
     {
-        $this->request = $request;
+        $this->request = $requestStack->getCurrentRequest();
```
Comment (Ener-Getick) [inline Search/QueryFactory.php]: should be implemented as described earlier

File: Resources/translations/FOSMessageBundle.cs.yml [hunk 13]
```diff
@@ -0,0 +1,27 @@
+messenger:      Zprávy
```
Comment (Ener-Getick) [inline Resources/translations/FOSMessageBundle.cs.yml]: Unrelated, can't we do that in another pr ?

File: Resources/config/routing.xml [hunk 14]
```diff
@@ -4,37 +4,37 @@
...
     
-    <route id="fos_message_thread_undelete" pattern="/{threadId}/undelete">
+    <route id="fos_message_thread_undelete" path="/{threadId}/undelete">
         <default key="_controller">FOSMessageBundle:Message:undelete</default>
-        <requirement key="_method">POST</requirement>
+        <requirement key="methods">POST</requirement>
     </route>
 
-    <route id="fos_message_thread_view" pattern="/{threadId}">
+    <route id="fos_message_thread_view" path="/{threadId}">
```
Comment (Ener-Getick) [inline Resources/config/routing.xml]: requires ``symfony/routing: ^2.2``

File: Resources/config/form.xml [hunk 15]
```diff
@@ -6,42 +6,38 @@
...
-        <service id="fos_message.new_thread_form.type.default" class="FOS\MessageBundle\FormType\NewThreadMessageFormType" public="false" />
-
         <service id="fos_message.new_thread_multiple_form.type" class="FOS\MessageBundle\FormType\NewThreadMultipleMessageFormType" public="false" />
 
-        <service id="fos_message.reply_form.type.default" class="FOS\MessageBundle\FormType\ReplyMessageFormType" public="false" />
-
         <service id="fos_message.new_thread_form.factory.default" class="FOS\MessageBundle\FormFactory\NewThreadMessageFormFactory" public="false">
             <argument type="service" id="form.factory" />
-            <argument type="service" id="fos_message.new_thread_form.type" />
+            <argument>%fos_message.new_thread_form.class%</argument>
```
Comment (Ener-Getick) [inline Resources/config/form.xml]: Should be managed in a compiler pass

**Comment by Ener-Getick** [comment]
BTW it looks like changes to ``composer.json`` were lost. (see https://github.com/FriendsOfSymfony/FOSMessageBundle/pull/275/files)

**Comment by Ener-Getick** [comment]
@ownede are you still interrested to work on this PR ?

**Comment by ownede** [comment]
@Ener-Getick 
Thanks for the review!

Well, this PR contains merges of various authors' forks. 

I'm still interested into further development, but it may take time. If somebody is also interested, please fell free to push your changes to my fork. 

btw. It seems to work fine with Symfony 3.0 on PHP 5.6, if anyone wants to use it before public release 😜

**Comment by Ener-Getick** [comment]
Take all the time you need :)
I don't use this bundle so i can't dedicate it too much time but i can review the pr of whoever wants to work on that.

File: Resources/config/config.xml (hunk 10 above)
Comment (tgalopin) [inline Resources/config/config.xml]: I don't, 2.3 is still supported when we should as well. Let's keep it this way and not throwing a depreciation yet.

Comment (Ener-Getick) [inline Resources/config/config.xml]: 2.3 only receives security fixes and it's really easy to upgrade to a higher version. I don't think we should give people an argument to not upgrade.

Comment (tgalopin) [inline Resources/config/config.xml]: Nevermind, that's wrong, 2.3 is not supported anymore. Let's drop it yes.

**Comment by tgalopin** [comment]
This should be fixed by https://github.com/FriendsOfSymfony/FOSMessageBundle/issues/289, I close but if you have problems, don't hesitate to open an issue :) !
